service/internal/access/v2/pdp.go

```diff
@@ -0,0 +1,299 @@
+	allAttributesByDefinitionFQN := make(map[string]*policy.Attribute)
+	allEntitleableAttributesByValueFQN := make(map[string]*attrs.GetAttributeValuesByFqnsResponse_AttributeAndValue)
+	for _, attr := range allAttributeDefinitions {
+		if err := validateAttribute(attr); err != nil {
+			return nil, fmt.Errorf("invalid attribute definition: %w", err)
+		}
+		allAttributesByDefinitionFQN[attr.GetFqn()] = attr
+
+		// Not every value may have a subject mapping and be entitleable, but a lookup must still be possible
+		for _, value := range attr.GetValues() {
+			mapped := &attrs.GetAttributeValuesByFqnsResponse_AttributeAndValue{
+				Value:     value,
+				Attribute: attr,
+			}
+			allEntitleableAttributesByValueFQN[value.GetFqn()] = mapped
+		}
+	}
+
+	for _, sm := range allSubjectMappings {
+		if err := validateSubjectMapping(sm); err != nil {
+			l.WarnContext(ctx, "invalid subject mapping - skipping", slog.Any("error", err), slog.Any("subject mapping", sm))
+			continue
+		}
+		mappedValue := sm.GetAttributeValue()
+		mappedValueFQN := mappedValue.GetFqn()
+		if _, ok := allEntitleableAttributesByValueFQN[mappedValueFQN]; ok {
+			allEntitleableAttributesByValueFQN[mappedValueFQN].Value.SubjectMappings = append(allEntitleableAttributesByValueFQN[mappedValueFQN].Value.SubjectMappings, sm)
+			continue
+		}
+		// Take subject mapping's attribute value and its definition from memory
+		parentDefinition, err := getDefinition(mappedValueFQN, allAttributesByDefinitionFQN)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get attribute definition: %w", err)
+		}
+		mappedValue.SubjectMappings = []*policy.SubjectMapping{sm}
+		mapped := &attrs.GetAttributeValuesByFqnsResponse_AttributeAndValue{
+			Value:     mappedValue,
+			Attribute: parentDefinition,
+		}
+		allEntitleableAttributesByValueFQN[mappedValueFQN] = mapped
+	}
+
+	pdp := &PolicyDecisionPoint{
+		l,
+		allEntitleableAttributesByValueFQN,
+	}
+	return pdp, nil
+}
+
+// GetDecision evaluates the action on the resources for the entity and returns a decision.
+func (p *PolicyDecisionPoint) GetDecision(
+	ctx context.Context,
+	entityRepresentation *entityresolutionV2.EntityRepresentation,
+	action *policy.Action,
+	resources []*authz.Resource,
+) (*Decision, error) {
+	loggable := []any{
+		slog.String("entity ID", entityRepresentation.GetOriginalId()),
+		slog.String("action", action.GetName()),
+		slog.Int("resources total", len(resources)),
+	}
+	p.logger.DebugContext(ctx, "getting decision", loggable...)
+
+	if err := validateGetDecision(entityRepresentation, action, resources); err != nil {
+		return nil, err
+	}
+
+	// Filter all attributes down to only those that relevant to the entitlement decisioning of these specific resources
+	decisionableAttributes := make(map[string]*attrs.GetAttributeValuesByFqnsResponse_AttributeAndValue)
+
+	for idx, resource := range resources {
+		// Assign indexed ephemeral ID for resource if not already set
+		if resource.GetEphemeralId() == "" {
```

Comment:
Can you explain the purpose of this ID, it's not immediately obvious to me.

Reply:
Thank you for the explanation.  I only saw them used in logging too.  No concerns given that use case, but I did notice that these can be supplied externally as well.  I was concerned there may be some potential for a user to introduce a conflict here.

But if only used in logging, I don't see a risk.